include/zephyr/net/http/service.h

```diff
@@ -0,0 +1,165 @@
+		.resource = _resource,                                                             \
+		.detail = (_detail),                                                               \
+	}
+
+struct http_service_desc {
+	const char *host;
+	uint16_t *port;
+	void *detail;
+	size_t concurrent;
+	size_t backlog;
+	struct http_resource_desc *res_begin;
+	struct http_resource_desc *res_end;
+};
+
+#define __z_http_service_define(_name, _host, _port, _concurrent, _backlog, _detail, _res_begin,   \
+				_res_end)                                                          \
+	static const STRUCT_SECTION_ITERABLE(http_service_desc, _name) = {                         \
+		.host = _host,                                                                     \
+		.port = (uint16_t *)(_port),                                                       \
+		.detail = (void *)(_detail),                                                       \
+		.concurrent = (_concurrent),                                                       \
+		.backlog = (_backlog),                                                             \
+		.res_begin = (_res_begin),                                                         \
+		.res_end = (_res_end),                                                             \
+	}
+
+/**
+ * @brief Define an HTTP service without static resources.
+ *
+ * @note The @p _host parameter must be non-`NULL`. It is used to specify an IP address either in
+ * IPv4 or IPv6 format a fully-qualified hostname or a virtual host.
+ *
+ * @note The @p _port parameter must be non-`NULL`. It points to a location that specifies the port
+ * number to use for the service. If the specified port number is zero, then an ephemeral port
+ * number will be used and the actual port number assigned will be written back memory (if that
+ * memory is writeable).
+ *
+ * @param _name Name of the service.
+ * @param _host IP address or hostname associated with the service.
+ * @param[inout] _port Pointer to port associated with the service.
+ * @param _concurrent Maximum number of concurrent clients.
+ * @param _backlog Maximum number queued connections.
+ * @param _detail Implementation-specific detail associated with the service.
+ */
+#define HTTP_SERVICE_DEFINE_EMPTY(_name, _host, _port, _concurrent, _backlog, _detail)             \
+	__z_http_service_define(_name, _host, _port, _concurrent, _backlog, _detail, 0, 0)
+
+/**
+ * @brief Define an HTTP service with static resources.
+ *
+ * @note The @p _host parameter must be non-`NULL`. It is used to specify an IP address either in
+ * IPv4 or IPv6 format a fully-qualified hostname or a virtual host.
+ *
+ * @note The @p _port parameter must be non-`NULL`. It points to a location that specifies the port
+ * number to use for the service. If the specified port number is zero, then an ephemeral port
+ * number will be used and the actual port number assigned will be written back memory (if that
+ * memory is writeable).
```

Comment:
And if it's not writeable? Sounds like a potential source od problems (and ambiguity in behaviour).

Reply:
Then it would read as whatever value it was initialized with.

Reply:
Should be possible for an implementation to check if a memory region is RO, IIRC.